gitpod-io/gitpod
# [supervisor] Terminate reparented processes during shutdown

This PR provides the same SIGTERM/SIGKILL shutdown mechanism to background processes started in terminals, as we provide to the terminals themselves.

As a consequence, one can now start tasks like:
```YAML
tasks:
  - before: sudo docker-up &
    # convenience for non-prebuild workspaces, not necessary to make this termination mechanism work
    command: fg
```
and expect the Docker daemon to terminate gracefully on workspace shutdown (within the terminal shutdown time: 10 seconds).

This PR also slightly extends the supervisor terminal API and introduces annotations on terminals (user-defined metadata). We now also report the terminal shell's PID.

### How does it work
In bash, background processes get disowned when the shell is terminated, and subsequently reparented to their parent process (i.e. supervisor). During supervisor shutdown, we now make the reaper send sigterm to any process that gets reparented to supervisor and is still running. In addition, we explicitly SIGTERM all child processes prior to shutting down. This includes former bash background processes.

### How to test
0. (Optional) Enable Feature Preview ... this adds another test for #2664 
1. Start a workspace on this [test repo](https://github.com/csweichel/test-repo/tree/proper-shutdown): https://cw-supervisor-2654.staging.gitpod-dev.com/#https://github.com/csweichel/test-repo/tree/proper-shutdown
2. Stop the workspace and notice the "proper-shutdown.txt" file in the list of dirty files
